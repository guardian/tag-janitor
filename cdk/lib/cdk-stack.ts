import { Schedule } from "@aws-cdk/aws-events";
import { PolicyStatement } from "@aws-cdk/aws-iam";
import { Runtime } from "@aws-cdk/aws-lambda";
import type { App } from "@aws-cdk/core";
import { Duration } from "@aws-cdk/core";
import type { GuStackProps } from "@guardian/cdk/lib/constructs/core";
import {
  GuStack,
  GuStringParameter,
  GuSubnetListParameter,
  GuVpcParameter,
} from "@guardian/cdk/lib/constructs/core";
import { GuVpc } from "@guardian/cdk/lib/constructs/ec2";
import { GuLambdaFunction } from "@guardian/cdk/lib/constructs/lambda";

export class CdkStack extends GuStack {
  constructor(scope: App, id: string, props: GuStackProps) {
    super(scope, id, props);

    const parameters = {
      bucketName: new GuStringParameter(this, "BucketName", {
        description: "BucketName",
      }),
      topic: new GuStringParameter(this, "topicArn", {
        description: "The ARN of the SNS topic to send messages to",
      }),
      vpc: new GuVpcParameter(this, "vpcId", {
        description:
          "The VPC ID for the lambda to live in (this allows it to talk to Prism)",
      }), // TODO: Look this up in SSM https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-ssm-parameter-types
      subnets: new GuSubnetListParameter(this, "subnetIds", {
        description:
          "The subnet IDs for the lambda to live in (this allows it to talk to Prism)",
      }), // TODO: Look this up in SSM https://docs.aws.amazon.com/AWSCloudFormation/latest/UserGuide/parameters-section-structure.html#aws-ssm-parameter-types
      accountsAllowList: new GuStringParameter(this, "accountsAllowList", {
        description: "A comma separated list of account numbers to include",
      }),
      prismUrl: new GuStringParameter(this, "prismUrl", {
        description: "Base URL for Prism",
      }),
    };

    const lambdaFrequency = Duration.days(7);

    const tagJanitorLambda = new GuLambdaFunction(this, `${this.app}-lambda`, {
      handler: "dist/handler.handler",
      functionName: `${this.app}-${this.stage}`,
      runtime: Runtime.NODEJS_12_X,
      code: {
        bucket: parameters.bucketName.valueAsString,
        key: `${this.stack}/${this.stage}/${this.app}/${this.app}.zip`,
      },
      environment: {
        STAGE: this.stage,
        TOPIC_ARN: parameters.topic.valueAsString,
        ACCOUNTS_ALLOW_LIST: parameters.accountsAllowList.valueAsString,
        PRISM_URL: parameters.prismUrl.valueAsString,
      },
      description: "Lambda to notify about instances with missing tags",
      timeout: Duration.seconds(30),
      memorySize: 512,
      vpc: GuVpc.fromId(this, "vpc", parameters.vpc.valueAsString),
      vpcSubnets: {
        subnets: GuVpc.subnets(this, parameters.subnets.valueAsList),
      },
      rules: [
        {
          schedule: Schedule.rate(lambdaFrequency),
          description: `Run tag-janitor every ${lambdaFrequency.toHumanString()}`,
        },
      ],
    });

    tagJanitorLambda.addToRolePolicy(
      new PolicyStatement({
        actions: ["SNS:Publish"],
        resources: [parameters.topic.valueAsString],
      })
    );
  }
}
