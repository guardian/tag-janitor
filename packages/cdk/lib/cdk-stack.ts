import { Schedule } from '@aws-cdk/aws-events';
import { PolicyStatement } from '@aws-cdk/aws-iam';
import { Runtime, RuntimeFamily } from '@aws-cdk/aws-lambda';
import type { App } from '@aws-cdk/core';
import { Duration } from '@aws-cdk/core';
import type { GuStackProps } from '@guardian/cdk/lib/constructs/core';
import { GuStack, GuStringParameter } from '@guardian/cdk/lib/constructs/core';
import { GuVpc } from '@guardian/cdk/lib/constructs/ec2';
import { GuScheduledLambda } from '@guardian/cdk/lib/patterns/scheduled-lambda';

export class CdkStack extends GuStack {
	constructor(scope: App, id: string, props: GuStackProps) {
		super(scope, id, props);

		const app = 'tag-janitor';

		const parameters = {
			topic: new GuStringParameter(this, 'topicArn', {
				description: 'The ARN of the SNS topic to send messages to',
			}),
			accountsAllowList: new GuStringParameter(this, 'accountsAllowList', {
				description: 'A comma separated list of account numbers to include',
			}),
			prismUrl: new GuStringParameter(this, 'prismUrl', {
				description: 'Base URL for Prism',
			}),
		};

		const lambdaFrequency = Duration.days(7);

		// This is copied from the aws-cdk codebase
		// TODO use the enum once updated version of @guardian/cdk
		const node16 = new Runtime('nodejs16.x', RuntimeFamily.NODEJS, {
			supportsInlineCode: true,
		});

		const tagJanitorLambda = new GuScheduledLambda(this, `${app}-lambda`, {
			app: app,
			handler: 'index.handler',
			functionName: `${app}-${this.stage}`,
			runtime: node16,
			fileName: `${app}.zip`,
			environment: {
				STAGE: this.stage,
				TOPIC_ARN: parameters.topic.valueAsString,
				ACCOUNTS_ALLOW_LIST: parameters.accountsAllowList.valueAsString,
				PRISM_URL: parameters.prismUrl.valueAsString,
			},
			description: 'Lambda to notify about instances with missing tags',
			// This lambda needs access to the Deploy Tools VPC so that it can talk to Prism
			vpc: GuVpc.fromIdParameter(this, 'vpc'),
			vpcSubnets: {
				subnets: GuVpc.subnetsFromParameter(this),
			},
			rules: [{ schedule: Schedule.rate(lambdaFrequency) }],
			monitoringConfiguration: {
				alarmName: `High error % from ${app}-${this.stage}`,
				snsTopicName: 'devx-alerts',
				toleratedErrorPercentage: 99,
			},
		});

		tagJanitorLambda.addToRolePolicy(
			new PolicyStatement({
				actions: ['SNS:Publish'],
				resources: [parameters.topic.valueAsString],
			}),
		);
	}
}
