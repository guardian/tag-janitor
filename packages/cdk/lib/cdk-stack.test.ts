import { SynthUtils } from '@aws-cdk/assert';
import { App } from '@aws-cdk/core';
import { CdkStack } from './cdk-stack';

describe('The tag-janitor stack', () => {
	it('matches the snapshot', () => {
		const app = new App();
		const stack = new CdkStack(app, 'tag-janitor', {
			stack: 'deploy',
			stage: 'PROD',
		});

		expect(SynthUtils.toCloudFormation(stack)).toMatchSnapshot();
	});
});
