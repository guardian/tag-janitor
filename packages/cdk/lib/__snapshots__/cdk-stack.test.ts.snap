// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The tag-janitor stack matches the snapshot 1`] = `
Object {
  "Mappings": Object {
    "stagemapping": Object {
      "CODE": Object {
        "alarmActionsEnabled": false,
      },
      "PROD": Object {
        "alarmActionsEnabled": true,
      },
    },
  },
  "Parameters": Object {
    "DistributionBucketName": Object {
      "Default": "/account/services/artifact.bucket",
      "Description": "SSM parameter containing the S3 bucket name holding distribution artifacts",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
    "PrivateSubnets": Object {
      "Default": "/account/vpc/primary/subnets/private",
      "Description": "A list of private subnets",
      "Type": "AWS::SSM::Parameter::Value<List<AWS::EC2::Subnet::Id>>",
    },
    "Stage": Object {
      "AllowedValues": Array [
        "CODE",
        "PROD",
      ],
      "Default": "CODE",
      "Description": "Stage name",
      "Type": "String",
    },
    "VpcId": Object {
      "Default": "/account/vpc/primary/id",
      "Description": "Virtual Private Cloud to run EC2 instances within",
      "Type": "AWS::SSM::Parameter::Value<AWS::EC2::VPC::Id>",
    },
    "accountsAllowList": Object {
      "Description": "A comma separated list of account numbers to include",
      "Type": "String",
    },
    "prismUrl": Object {
      "Description": "Base URL for Prism",
      "Type": "String",
    },
    "topicArn": Object {
      "Description": "The ARN of the SNS topic to send messages to",
      "Type": "String",
    },
  },
  "Resources": Object {
    "ErrorPercentageAlarmForLambda54F48384": Object {
      "Properties": Object {
        "ActionsEnabled": Object {
          "Fn::FindInMap": Array [
            "stagemapping",
            Object {
              "Ref": "Stage",
            },
            "alarmActionsEnabled",
          ],
        },
        "AlarmActions": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:aws:sns:",
                Object {
                  "Ref": "AWS::Region",
                },
                ":",
                Object {
                  "Ref": "AWS::AccountId",
                },
                ":devx-alerts",
              ],
            ],
          },
        ],
        "AlarmDescription": Object {
          "Fn::Join": Array [
            "",
            Array [
              Object {
                "Ref": "tagjanitorlambda3E6E11A1",
              },
              " exceeded 99% error rate",
            ],
          ],
        },
        "AlarmName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "High error % from tag-janitor-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "ComparisonOperator": "GreaterThanThreshold",
        "EvaluationPeriods": 1,
        "Metrics": Array [
          Object {
            "Expression": "100*m1/m2",
            "Id": "expr_1",
            "Label": Object {
              "Fn::Join": Array [
                "",
                Array [
                  "Error % of ",
                  Object {
                    "Ref": "tagjanitorlambda3E6E11A1",
                  },
                ],
              ],
            },
          },
          Object {
            "Id": "m1",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "tagjanitorlambda3E6E11A1",
                    },
                  },
                ],
                "MetricName": "Errors",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
          Object {
            "Id": "m2",
            "MetricStat": Object {
              "Metric": Object {
                "Dimensions": Array [
                  Object {
                    "Name": "FunctionName",
                    "Value": Object {
                      "Ref": "tagjanitorlambda3E6E11A1",
                    },
                  },
                ],
                "MetricName": "Invocations",
                "Namespace": "AWS/Lambda",
              },
              "Period": 60,
              "Stat": "Sum",
            },
            "ReturnData": false,
          },
        ],
        "Threshold": 99,
        "TreatMissingData": "notBreaching",
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "tagjanitorlambda3E6E11A1": Object {
      "DependsOn": Array [
        "tagjanitorlambdaServiceRoleDefaultPolicy79766902",
        "tagjanitorlambdaServiceRoleD63E9AA7",
      ],
      "Properties": Object {
        "Code": Object {
          "S3Bucket": Object {
            "Ref": "DistributionBucketName",
          },
          "S3Key": Object {
            "Fn::Join": Array [
              "",
              Array [
                "deploy/",
                Object {
                  "Ref": "Stage",
                },
                "/tag-janitor/tag-janitor.zip",
              ],
            ],
          },
        },
        "Description": "Lambda to notify about instances with missing tags",
        "Environment": Object {
          "Variables": Object {
            "ACCOUNTS_ALLOW_LIST": Object {
              "Ref": "accountsAllowList",
            },
            "APP": "tag-janitor",
            "PRISM_URL": Object {
              "Ref": "prismUrl",
            },
            "STACK": "deploy",
            "STAGE": Object {
              "Ref": "Stage",
            },
            "TOPIC_ARN": Object {
              "Ref": "topicArn",
            },
          },
        },
        "FunctionName": Object {
          "Fn::Join": Array [
            "",
            Array [
              "tag-janitor-",
              Object {
                "Ref": "Stage",
              },
            ],
          ],
        },
        "Handler": "src/handler.handler",
        "MemorySize": 512,
        "Role": Object {
          "Fn::GetAtt": Array [
            "tagjanitorlambdaServiceRoleD63E9AA7",
            "Arn",
          ],
        },
        "Runtime": "nodejs16.x",
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "tag-janitor",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/tag-janitor",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "Timeout": 30,
        "VpcConfig": Object {
          "SecurityGroupIds": Array [
            Object {
              "Fn::GetAtt": Array [
                "tagjanitorlambdaSecurityGroupCDEEB485",
                "GroupId",
              ],
            },
          ],
          "SubnetIds": Object {
            "Ref": "PrivateSubnets",
          },
        },
      },
      "Type": "AWS::Lambda::Function",
    },
    "tagjanitorlambdaSecurityGroupCDEEB485": Object {
      "Properties": Object {
        "GroupDescription": "Automatic security group for Lambda Function tagjanitortagjanitorlambdaBA439B3D",
        "SecurityGroupEgress": Array [
          Object {
            "CidrIp": "0.0.0.0/0",
            "Description": "Allow all outbound traffic by default",
            "IpProtocol": "-1",
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "tag-janitor",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/tag-janitor",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
        "VpcId": Object {
          "Ref": "VpcId",
        },
      },
      "Type": "AWS::EC2::SecurityGroup",
    },
    "tagjanitorlambdaServiceRoleD63E9AA7": Object {
      "Properties": Object {
        "AssumeRolePolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": Object {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": Array [
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
          Object {
            "Fn::Join": Array [
              "",
              Array [
                "arn:",
                Object {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaVPCAccessExecutionRole",
              ],
            ],
          },
        ],
        "Tags": Array [
          Object {
            "Key": "App",
            "Value": "tag-janitor",
          },
          Object {
            "Key": "gu:cdk:version",
            "Value": "TEST",
          },
          Object {
            "Key": "gu:repo",
            "Value": "guardian/tag-janitor",
          },
          Object {
            "Key": "Stack",
            "Value": "deploy",
          },
          Object {
            "Key": "Stage",
            "Value": Object {
              "Ref": "Stage",
            },
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "tagjanitorlambdaServiceRoleDefaultPolicy79766902": Object {
      "Properties": Object {
        "PolicyDocument": Object {
          "Statement": Array [
            Object {
              "Action": Array [
                "s3:GetObject*",
                "s3:GetBucket*",
                "s3:List*",
              ],
              "Effect": "Allow",
              "Resource": Array [
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                    ],
                  ],
                },
                Object {
                  "Fn::Join": Array [
                    "",
                    Array [
                      "arn:",
                      Object {
                        "Ref": "AWS::Partition",
                      },
                      ":s3:::",
                      Object {
                        "Ref": "DistributionBucketName",
                      },
                      "/*",
                    ],
                  ],
                },
              ],
            },
            Object {
              "Action": "ssm:GetParametersByPath",
              "Effect": "Allow",
              "Resource": Object {
                "Fn::Join": Array [
                  "",
                  Array [
                    "arn:aws:ssm:",
                    Object {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    Object {
                      "Ref": "AWS::AccountId",
                    },
                    ":parameter/",
                    Object {
                      "Ref": "Stage",
                    },
                    "/deploy/tag-janitor",
                  ],
                ],
              },
            },
            Object {
              "Action": "SNS:Publish",
              "Effect": "Allow",
              "Resource": Object {
                "Ref": "topicArn",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "tagjanitorlambdaServiceRoleDefaultPolicy79766902",
        "Roles": Array [
          Object {
            "Ref": "tagjanitorlambdaServiceRoleD63E9AA7",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "tagjanitorlambdatagjanitorlambdarate7days0AllowEventRuletagjanitortagjanitorlambdaBA439B3DF4A71923": Object {
      "Properties": Object {
        "Action": "lambda:InvokeFunction",
        "FunctionName": Object {
          "Fn::GetAtt": Array [
            "tagjanitorlambda3E6E11A1",
            "Arn",
          ],
        },
        "Principal": "events.amazonaws.com",
        "SourceArn": Object {
          "Fn::GetAtt": Array [
            "tagjanitorlambdatagjanitorlambdarate7days0B5F68379",
            "Arn",
          ],
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "tagjanitorlambdatagjanitorlambdarate7days0B5F68379": Object {
      "Properties": Object {
        "ScheduleExpression": "rate(7 days)",
        "State": "ENABLED",
        "Targets": Array [
          Object {
            "Arn": Object {
              "Fn::GetAtt": Array [
                "tagjanitorlambda3E6E11A1",
                "Arn",
              ],
            },
            "Id": "Target0",
          },
        ],
      },
      "Type": "AWS::Events::Rule",
    },
  },
}
`;
